// @ghom/logger/dist/cjs/app/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.defaultLoggerRenders = exports.defaultLoggerColors = exports.defaultLoggerPattern = exports.formatLoggerLevelName = exports.loggerLevelName = exports.LoggerLevels = void 0;
const chalk_1 = __importDefault(require("chalk"));
const dayjs_1 = __importDefault(require("dayjs"));
var LoggerLevels;
(function (LoggerLevels) {
    LoggerLevels[LoggerLevels["WARN"] = 0] = "WARN";
    LoggerLevels[LoggerLevels["ERROR"] = 1] = "ERROR";
    LoggerLevels[LoggerLevels["INFO"] = 2] = "INFO";
    LoggerLevels[LoggerLevels["SUCCESS"] = 3] = "SUCCESS";
})(LoggerLevels = exports.LoggerLevels || (exports.LoggerLevels = {}));
function loggerLevelName(level) {
    return Object.keys(LoggerLevels).find((key) => LoggerLevels[key] === level);
}
exports.loggerLevelName = loggerLevelName;
function formatLoggerLevelName(levelName) {
    const maxLength = Math.max(...Object.keys(LoggerLevels).map((key) => key.length));
    return levelName.padStart(maxLength, " ");
}
exports.formatLoggerLevelName = formatLoggerLevelName;
const defaultLoggerPattern = (text, config, secondaryText) => {
    return `${chalk_1.default.grey((0, dayjs_1.default)().format("DD/MM/YY HH:mm"))} ${chalk_1.default.hex(config.colors[config.level])(formatLoggerLevelName(loggerLevelName(config.level)))}${secondaryText
        ? " " + chalk_1.default.hex(config.colors.secondaryText)(`${secondaryText}`)
        : ""} ${text}`;
};
exports.defaultLoggerPattern = defaultLoggerPattern;
exports.defaultLoggerColors = {
    [LoggerLevels.WARN]: "#ffa600",
    [LoggerLevels.ERROR]: "#ff0000",
    [LoggerLevels.INFO]: "#00ffff",
    [LoggerLevels.SUCCESS]: "#00ff00",
    secondaryText: "#ff3fa8",
};
exports.defaultLoggerRenders = {
    [LoggerLevels.WARN]: console.warn,
    [LoggerLevels.ERROR]: console.error,
    [LoggerLevels.INFO]: console.log,
    [LoggerLevels.SUCCESS]: console.log,
};
class Logger {
    constructor(options) {
        this.section = options?.section;
        this.pattern = options?.pattern ?? exports.defaultLoggerPattern;
        this.colors = options?.colors ?? exports.defaultLoggerColors;
        this.renders = options?.renders ?? exports.defaultLoggerRenders;
    }
    log(text) {
        this.renders[LoggerLevels.INFO](this.pattern(text, { level: LoggerLevels.INFO, colors: this.colors }, this.section));
    }
    error(text, _path, full) {
        this.renders[LoggerLevels.ERROR](this.pattern(text instanceof Error ? text.message : text, { level: LoggerLevels.ERROR, colors: this.colors }, _path ?? this.section));
        if (full && text instanceof Error)
            this.renders[LoggerLevels.ERROR](text);
    }
    warn(text) {
        this.renders[LoggerLevels.WARN](this.pattern(text, { level: LoggerLevels.WARN, colors: this.colors }, this.section));
    }
    success(text) {
        this.renders[LoggerLevels.SUCCESS](this.pattern(text, { level: LoggerLevels.SUCCESS, colors: this.colors }, this.section));
    }
}
exports.Logger = Logger;
